File: backend/src/main/java/com/thesis/ecoursemanagement/service/CourseService.java
package com.thesis.ecoursemanagement.service;

import com.thesis.ecoursemanagement.dto.request.CourseRequest;
import com.thesis.ecoursemanagement.dto.response.ClassResponse;
import com.thesis.ecoursemanagement.dto.response.CourseResponse;
import com.thesis.ecoursemanagement.mapper.ClassMapper;
import com.thesis.ecoursemanagement.mapper.CourseMapper;
import com.thesis.ecoursemanagement.model.Course;
import com.thesis.ecoursemanagement.repository.CourseRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class CourseService {
    private final CourseMapper courseMapper;
    private final CourseRepository courseRepository;
    private final ClassMapper classMapper;

    public List<CourseResponse> getAllCourses() {
        return courseMapper.toResponseList(courseRepository.findAll());
    }

    public CourseResponse getCourseById(Long id) {
        Course course = courseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Course not found"));

        List<ClassResponse> classResponses = classMapper.toResponseList(course.getClasses());

        return CourseResponse.builder()
                .id(course.getId())
                .courseName(course.getCourseName())
                .description(course.getDescription())
                .classes(classResponses)
                .build();
    }

    public CourseResponse createCourse(CourseRequest request) {
        Course course = courseMapper.toCourse(request);
        return courseMapper.toResponse(courseRepository.save(course));
    }

    public CourseResponse updateCourse(Long id, CourseRequest request) {
        Course course = courseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Course not found"));
        if (request.getCourseName() != null) course.setCourseName(request.getCourseName());
        if (request.getDescription() != null) course.setDescription(request.getDescription());
        return courseMapper.toResponse(courseRepository.save(course));
    }

    public void deleteCourse(Long id) {
        courseRepository.deleteById(id);
    }
}

File: frontend/ecourse-management/src/services/courseService.js
import api from "./apis.js";


export const getAllCourses = async () => {
    try {
        const response = await api.get("/courses");
        return response.data;
    } catch (error) {
        console.error("Lỗi khi lấy danh sách khóa học:", error);
        throw error;
    }
};

export const getCourseById = async (id) => {
    try {
        const response = await api.get(`/courses/${id}`);
        return response.data;
    } catch (error) {
        console.error("Lỗi khi lấy chi tiết khóa học:", error);
        throw error;
    }
};

export const deleteCourse = async (courseId) => {
    try {
        const response = await api.delete(`/courses/${courseId}`);
        return response.data;
    } catch (error) { }
};

export const createCourse = async ({ courseName, description }) => {
    try {
        const response = await api.post("/courses", { courseName, description });
        return response.data;
    } catch (error) {
        console.error("Lỗi khi tạo khóa học:", error);
        throw error;
    }
};

export const updateCourse = async (courseId, courseName, description) => {
    try {
        const response = await api.put(`/courses/${courseId}`, { courseName, description });
        return response.data;
    } catch (error) {
        console.error("Lỗi khi cập nhật khóa học:", error);
        throw error;
    }
};
